ASSERTION FAILED: !isMainRunLoop() in WebCore::SWRegistrationDatabase::close

fujii opened the pull request:
#### ee0c81c806576cd1469fe3ea65a4154f8700399b
<pre>
ASSERTION FAILED: !isMainRunLoop() in WebCore::SWRegistrationDatabase::close
<a href="https://bugs.webkit.org/show_bug.cgi?id=258000">https://bugs.webkit.org/show_bug.cgi?id=258000</a>

Reviewed by Youenn Fablet.

NetworkStorageManager should destroy
m_sharedServiceWorkerStorageManager in the m_queue thread. Added
WTF_GUARDED_BY_CAPABILITY for m_sharedServiceWorkerStorageManager.
Added more assertIsCurrent(workQueue()) in lambda functions.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::close):
(WebKit::NetworkStorageManager::clearServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::importServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::updateServiceWorkerRegistrations):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:

Canonical link: <a href="https://commits.webkit.org/265176@main">https://commits.webkit.org/265176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5cf15ff277b9bd79c7f992a220cd035779d8a97

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9875 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10123 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10371 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11528 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9580 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12108 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10075 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12526 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10029 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10832 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11912 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8138 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16301 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9237 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9107 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12382 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9607 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7807 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8769 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2423 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12993 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9382 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -222,6 +222,7 @@ NetworkStorageManager::~NetworkStorageManager()
 {
     ASSERT(RunLoop::isMain());
     ASSERT(m_closed);
+    m_queue->dispatch([storageManger = WTFMove(m_sharedServiceWorkerStorageManager)] { });
```
Good catch! This should be moved to `NetworkStorageManager::close`, inside the task dispatched to m_queue.

fujii replied:
Good point. Will fix. However, doing `m_sharedServiceWorkerStorageManager = nullptr;` in the dispatched task causes another problem if I put a break point in `SWRegistrationDatabase::close`. I have to use `WTFMove(m_sharedServiceWorkerStorageManager)` in the lambda capture.

fujii replied:
Fixed.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -253,7 +253,7 @@ void NetworkStorageManager::close(CompletionHandler<void()>&& completionHandler)
     });
 
-    m_queue->dispatch([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)]() mutable {
+    m_queue->dispatch([this, protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler), storageManager = WTFMove(m_sharedServiceWorkerStorageManager)]() mutable {
```
m_sharedServiceWorkerStorageManager is set in the work queue, so it could be moved here, but set later on to a non null value.
It might be good to:
- Add a WTF_GUARDED_BY_CAPABILITY to m_sharedServiceWorkerStorageManager
- If setting m_sharedServiceWorkerStorageManager to null in the lambda does not work, std::exchange should work here.

fujii replied:
Good point. Fixed.

fujii committed:
ASSERTION FAILED: !isMainRunLoop() in WebCore::SWRegistrationDatabase::close
https://bugs.webkit.org/show_bug.cgi?id=258000

Reviewed by Youenn Fablet.

NetworkStorageManager should destroy
m_sharedServiceWorkerStorageManager in the m_queue thread. Added
WTF_GUARDED_BY_CAPABILITY for m_sharedServiceWorkerStorageManager.
Added more assertIsCurrent(workQueue()) in lambda functions.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::close):
(WebKit::NetworkStorageManager::clearServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::importServiceWorkerRegistrations):
(WebKit::NetworkStorageManager::updateServiceWorkerRegistrations):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:

Canonical link: https://commits.webkit.org/265176@main
